### ISSUE TITLE
Price part crashes if default currency has never been set.

### ISSUE THREAD

**Issue by bleroy** [OrchardCMS/OrchardCore.Commerce#35]
Stack trace:

```
2019-05-25 23:32:25.7544|Default|80000024-0006-fe00-b63f-84710c7967bb||OrchardCore.ContentManagement.Display.ContentItemDisplayCoordinator|ERROR|IContentPartDisplayDriver thrown from OrchardCore.Commerce.Drivers.PricePartDisplayDriver by ArgumentOutOfRangeException System.ArgumentOutOfRangeException: Currency not found.
Parameter name: isoSymbol
   at OrchardCore.Commerce.Services.MoneyService.GetCurrency(String isoSymbol) in C:\Projects\OrchardCore\src\OrchardCore.Modules\OrchardCore.Commerce\Services\MoneyService.cs:line 39
   at OrchardCore.Commerce.Services.MoneyService.get_DefaultCurrency() in C:\Projects\OrchardCore\src\OrchardCore.Modules\OrchardCore.Commerce\Services\MoneyService.cs:line 29
   at OrchardCore.Commerce.Drivers.PricePartDisplayDriver.BuildViewModel(PricePartViewModel model, PricePart part) in C:\Projects\OrchardCore\src\OrchardCore.Modules\OrchardCore.Commerce\Drivers\PricePartDisplayDriver.cs:line 51
   at OrchardCore.Commerce.Drivers.PricePartDisplayDriver.<>c__DisplayClass3_0.<Edit>b__0(PricePartViewModel m) in C:\Projects\OrchardCore\src\OrchardCore.Modules\OrchardCore.Commerce\Drivers\PricePartDisplayDriver.cs:line 32
   at OrchardCore.DisplayManagement.ShapeFactoryExtensions.<>c__DisplayClass6_0`1.<<CreateAsync>b__0>d.MoveNext() in C:\Projects\OrchardCore\src\OrchardCore\OrchardCore.DisplayManagement\IShapeFactory.cs:line 76
--- End of stack trace from previous location where exception was thrown ---
   at OrchardCore.DisplayManagement.Implementation.DefaultShapeFactory.CreateAsync(String shapeType, Func`1 shapeFactory, Action`1 creating, Action`1 created) in C:\Projects\OrchardCore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultShapeFactory.cs:line 91
   at OrchardCore.DisplayManagement.Views.ShapeResult.ApplyImplementationAsync(BuildShapeContext context, String displayType) in C:\Projects\OrchardCore\src\OrchardCore\OrchardCore.DisplayManagement\Views\ShapeResult.cs:line 106
   at OrchardCore.ContentManagement.Display.ContentItemDisplayCoordinator.<>c__DisplayClass11_2.<<BuildEditorAsync>b__0>d.MoveNext() in C:\Projects\OrchardCore\src\OrchardCore\OrchardCore.ContentManagement.Display\ContentDisplay\ContentItemDisplayCoordinator.cs:line 203
--- End of stack trace from previous location where exception was thrown ---
   at OrchardCore.Modules.InvokeExtensions.InvokeAsync[TEvents](IEnumerable`1 events, Func`2 dispatch, ILogger logger) in C:\Projects\OrchardCore\src\OrchardCore\OrchardCore.Abstractions\Modules\Extensions\InvokeExtensions.cs:line 78    at OrchardCore.Commerce.Services.MoneyService.GetCurrency(String isoSymbol) in C:\Projects\OrchardCore\src\OrchardCore.Modules\OrchardCore.Commerce\Services\MoneyService.cs:line 39
   at OrchardCore.Commerce.Services.MoneyService.get_DefaultCurrency() in C:\Projects\OrchardCore\src\OrchardCore.Modules\OrchardCore.Commerce\Services\MoneyService.cs:line 29
   at OrchardCore.Commerce.Drivers.PricePartDisplayDriver.BuildViewModel(PricePartViewModel model, PricePart part) in C:\Projects\OrchardCore\src\OrchardCore.Modules\OrchardCore.Commerce\Drivers\PricePartDisplayDriver.cs:line 51
   at OrchardCore.Commerce.Drivers.PricePartDisplayDriver.<>c__DisplayClass3_0.<Edit>b__0(PricePartViewModel m) in C:\Projects\OrchardCore\src\OrchardCore.Modules\OrchardCore.Commerce\Drivers\PricePartDisplayDriver.cs:line 32
   at OrchardCore.DisplayManagement.ShapeFactoryExtensions.<>c__DisplayClass6_0`1.<<CreateAsync>b__0>d.MoveNext() in C:\Projects\OrchardCore\src\OrchardCore\OrchardCore.DisplayManagement\IShapeFactory.cs:line 76
--- End of stack trace from previous location where exception was thrown ---
   at OrchardCore.DisplayManagement.Implementation.DefaultShapeFactory.CreateAsync(String shapeType, Func`1 shapeFactory, Action`1 creating, Action`1 created) in C:\Projects\OrchardCore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultShapeFactory.cs:line 91
   at OrchardCore.DisplayManagement.Views.ShapeResult.ApplyImplementationAsync(BuildShapeContext context, String displayType) in C:\Projects\OrchardCore\src\OrchardCore\OrchardCore.DisplayManagement\Views\ShapeResult.cs:line 106
   at OrchardCore.ContentManagement.Display.ContentItemDisplayCoordinator.<>c__DisplayClass11_2.<<BuildEditorAsync>b__0>d.MoveNext() in C:\Projects\OrchardCore\src\OrchardCore\OrchardCore.ContentManagement.Display\ContentDisplay\ContentItemDisplayCoordinator.cs:line 203
--- End of stack trace from previous location where exception was thrown ---
   at OrchardCore.Modules.InvokeExtensions.InvokeAsync[TEvents](IEnumerable`1 events, Func`2 dispatch, ILogger logger) in C:\Projects\OrchardCore\src\OrchardCore\OrchardCore.Abstractions\Modules\Extensions\InvokeExtensions.cs:line 78
```

**Comment by jeffolmstead** [comment]
I still encountered this issue on a deployment I just did. I see I am targeting 2.2.0 though and and in your fix you aren't. Is that a factor in the "$" not being recognized? Everything works when I manually change the setting to USD.

**Comment by bleroy** [comment]
I'll upgrade and check, but in the meantime, if you have a stack trace, that would probably help as well.

**Comment by jeffolmstead** [comment]
Here is stack trace. The environment was using Nuget/MyGet with NET 2.2.0. 

```
2019-09-10 06:56:27.0602|BRVN|0HLPLRHUUNTLD:00000020||OrchardCore.ContentManagement.Display.ContentItemDisplayCoordinator|ERROR|IContentPartDisplayDriver thrown from OrchardCore.Commerce.Drivers.PricePartDisplayDriver by ArgumentOutOfRangeException System.ArgumentOutOfRangeException: Currency not found.
Parameter name: isoSymbol
Actual value was $.
   at OrchardCore.Commerce.Services.MoneyService.GetCurrency(String isoSymbol) in C:\Users\Jeffrey\source\repos\EMS.Hosting\EMS.Hosting\Modules\OrchardCore.Commerce\Services\MoneyService.cs:line 47
   at OrchardCore.Commerce.Services.MoneyService.get_DefaultCurrency() in C:\Users\Jeffrey\source\repos\EMS.Hosting\EMS.Hosting\Modules\OrchardCore.Commerce\Services\MoneyService.cs:line 35
   at OrchardCore.Commerce.Drivers.PricePartDisplayDriver.BuildViewModel(PricePartViewModel model, PricePart part) in C:\Users\Jeffrey\source\repos\EMS.Hosting\EMS.Hosting\Modules\OrchardCore.Commerce\Drivers\PricePartDisplayDriver.cs:line 51
   at OrchardCore.Commerce.Drivers.PricePartDisplayDriver.<>c__DisplayClass3_0.<Edit>b__0(PricePartViewModel m) in C:\Users\Jeffrey\source\repos\EMS.Hosting\EMS.Hosting\Modules\OrchardCore.Commerce\Drivers\PricePartDisplayDriver.cs:line 32
   at OrchardCore.DisplayManagement.ShapeFactoryExtensions.<>c__DisplayClass6_0`1.<<CreateAsync>b__0>d.MoveNext() in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\IShapeFactory.cs:line 78
--- End of stack trace from previous location where exception was thrown ---
   at OrchardCore.DisplayManagement.Implementation.DefaultShapeFactory.CreateAsync(String shapeType, Func`1 shapeFactory, Action`1 creating, Action`1 created) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultShapeFactory.cs:line 138
   at OrchardCore.DisplayManagement.Views.ShapeResult.ApplyImplementationAsync(BuildShapeContext context, String displayType) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Views\ShapeResult.cs:line 199
   at OrchardCore.ContentManagement.Display.ContentItemDisplayCoordinator.<>c__DisplayClass11_2.<<BuildEditorAsync>b__0>d.MoveNext() in C:\projects\orchardcore\src\OrchardCore\OrchardCore.ContentManagement.Display\ContentDisplay\ContentItemDisplayCoordinator.cs:line 209
--- End of stack trace from previous location where exception was thrown ---
   at OrchardCore.Modules.InvokeExtensions.InvokeAsync[TEvents](IEnumerable`1 events, Func`2 dispatch, ILogger logger) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.Abstractions\Modules\Extensions\InvokeExtensions.cs:line 85    at OrchardCore.Commerce.Services.MoneyService.GetCurrency(String isoSymbol) in C:\Users\Jeffrey\source\repos\EMS.Hosting\EMS.Hosting\Modules\OrchardCore.Commerce\Services\MoneyService.cs:line 47
   at OrchardCore.Commerce.Services.MoneyService.get_DefaultCurrency() in C:\Users\Jeffrey\source\repos\EMS.Hosting\EMS.Hosting\Modules\OrchardCore.Commerce\Services\MoneyService.cs:line 35
   at OrchardCore.Commerce.Drivers.PricePartDisplayDriver.BuildViewModel(PricePartViewModel model, PricePart part) in C:\Users\Jeffrey\source\repos\EMS.Hosting\EMS.Hosting\Modules\OrchardCore.Commerce\Drivers\PricePartDisplayDriver.cs:line 51
   at OrchardCore.Commerce.Drivers.PricePartDisplayDriver.<>c__DisplayClass3_0.<Edit>b__0(PricePartViewModel m) in C:\Users\Jeffrey\source\repos\EMS.Hosting\EMS.Hosting\Modules\OrchardCore.Commerce\Drivers\PricePartDisplayDriver.cs:line 32
   at OrchardCore.DisplayManagement.ShapeFactoryExtensions.<>c__DisplayClass6_0`1.<<CreateAsync>b__0>d.MoveNext() in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\IShapeFactory.cs:line 78
--- End of stack trace from previous location where exception was thrown ---
   at OrchardCore.DisplayManagement.Implementation.DefaultShapeFactory.CreateAsync(String shapeType, Func`1 shapeFactory, Action`1 creating, Action`1 created) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Implementation\DefaultShapeFactory.cs:line 138
   at OrchardCore.DisplayManagement.Views.ShapeResult.ApplyImplementationAsync(BuildShapeContext context, String displayType) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.DisplayManagement\Views\ShapeResult.cs:line 199
   at OrchardCore.ContentManagement.Display.ContentItemDisplayCoordinator.<>c__DisplayClass11_2.<<BuildEditorAsync>b__0>d.MoveNext() in C:\projects\orchardcore\src\OrchardCore\OrchardCore.ContentManagement.Display\ContentDisplay\ContentItemDisplayCoordinator.cs:line 209
--- End of stack trace from previous location where exception was thrown ---
   at OrchardCore.Modules.InvokeExtensions.InvokeAsync[TEvents](IEnumerable`1 events, Func`2 dispatch, ILogger logger) in C:\projects\orchardcore\src\OrchardCore\OrchardCore.Abstractions\Modules\Extensions\InvokeExtensions.cs:line 85
```

**Comment by jeffolmstead** [comment]
-- Note: Moved comment here from code fix as more relevant to this conversation

In MoneyServices.cs in "GetCurrency" if you change code from this:

```
Currencies.FirstOrDefault(p => p.IsoCode.Equals(isoSymbol, StringComparison.InvariantCultureIgnoreCase))
```

to this:

```
Currencies.FirstOrDefault(p => p.Symbol.Equals(isoSymbol, StringComparison.InvariantCultureIgnoreCase))
```

Then it works the first time. **But**, after you actually set the default it passes in "USD" so then this would break. I think the confusion is that the incoming parameter is called "isoSymbol" but really what you want to pass in is "IsoCode". 

With this in mind, I believe the issue comes from CommerceSettingsConfiguration.cs in "Configure" where the DefaultCurrency is set to a "symbol" (e.g. "$" instead of "USD") if no default currency is found. If I had to guess, the methods started out using the symbol but then the architecture moved to using the ISO Code. If that is the case, then this change here should fix it:

Change from:
```
options.DefaultCurrency =
    String.IsNullOrEmpty(defaultCurrencySymbol) ? Currency.Dollar.Symbol : defaultCurrencySymbol;
```

To:
```
options.DefaultCurrency =
    String.IsNullOrEmpty(defaultCurrencySymbol) ? Currency.Dollar.IsoCode : defaultCurrencySymbol;
```

Hopefully this is helpful.
